arangod/Cluster/AgencyCache.cpp

```diff
@@ -92,6 +93,43 @@ index_t AgencyCache::index() const {
   return _commitIndex;
 }
 
+
+void AgencyCache::handleCallbacksNoLock(VPackSlice slice, std::unordered_set<uint64_t>& toCall) {
+
+  if (!slice.isObject()) {
+    LOG_TOPIC("31514", DEBUG, Logger::CLUSTER) <<
+      "Cannot handle callback on non-object " << slice.toJson();
+    return;
+  }
+
+  // Collect and normalize keys
+  std::vector<std::string> keys;
+  keys.reserve(slice.length());
+  for (auto const& i : VPackObjectIterator(slice)) {
+    keys.push_back(Node::normalize(i.key.copyString()));
+  }
+  std::sort(keys.begin(), keys.end());
+
+  for (auto const& cb : _callbacks) {
+    std::vector<std::string>::iterator pos = keys.begin();
+    do {
+      pos = std::find_if(
```

Comment:
This is quadratic, because we do a linear search for each callback. This would be problematic if we have say 1000 callbacks and a large transaction with 10000 changes. I thought we wanted to do something like this:

```
for cb : _callbacks {
  it = lower_bound(keys, cb);
  check if cb.key is a prefix of *it, if so add callback to set
}
for (k : keys) {
  it = _callbacks.lower_bound(k);
  while (it != end() && it->key is prefix of k) {
    add *it to callback set
    ++it;
  }
}
```